### ISSUE TITLE
Source strings are not correctly detected as UTF-8

### ISSUE THREAD

**Issue by mgruner** [bibletime/bibletime#86]
In `src/backend/managers/clanguagemgr.cpp`, there are a number of strings containing Unicode escape sequences (`\u`). This is displayed incorrectly in the application (at least if a translation is used), and lands broken in the language translation template file.

```
m_langList.append( new Language("itz", "Itz\u00e1", QObject::tr("Itz\u00e1")) );
```
becomes:
```
    <message>
        <location filename="../../src/backend/managers/clanguagemgr.cpp" line="292"/>
        <source>Itzu00e1</source>
        <extracomment>Language name itz</extracomment>
        <translation type="unfinished"></translation>
    </message>
´´´

This can obviously not be translated correctly.

**Comment by jaakristioja** [comment]
This is likely caused by Qt's `lupdate` tool, which does not correctly parse the `\u` Unicode escapes. If we prefix these strings with u8 in our C++ code, `make messages` (i.e. Qt's `lupdate`) will not properly detect these strings at all. I doubt there's a good way to work around the `lupdate` issue from our side. I haven't yet searched whether there's an upstream bug report for this.

**Comment by jaakristioja** [comment]
> This is displayed incorrectly in the application (at least if a translation is used)

Where exactly can I find these strings in the BibleTime application?

**Comment by mgruner** [comment]
They are defined in clanguagemgr.cpp. Probably to translate them for modules showing up in the Bookshelf manager.
Translation sources are defined in i18n/messages/bibletime_ui.ts.

**Comment by mgruner** [comment]
Yes, if you open the Booshelf manager with modules from a Wycliffe repository, you should see these strings.

**Comment by mgruner** [comment]
You need to select grouping by language.

**Comment by gholmlund** [comment]
https://bugreports.qt.io/browse/QTBUG-35164
I added a vote for this. I don't know what we can do about it.

**Comment by jaakristioja** [comment]
In principle, for a workaround, maybe we could write a script which after `lupdate` scans the generated `*.ts` files, and for all strings containing a `u` character followed by at least 4 hex digits, looks up the corresponding string in the linked source file and checks whether it contains any `\u` escapes. If it does, it fixes the respective string in the translation. The translation files have `<?xml version="1.0" encoding="utf-8"?>` so I think this could work.

**Comment by jaakristioja** [comment]
Here's a quick and extremely ugly hack I put together to fix a single translation file. Compiles, runs, seems to do something. Haven't completely verified the output, nor done much testing. It's 3 A.M. here and my English skills are failing, so I try not write moar wordz. gn

```C++
#include <cstddef>
#include <cstdlib>
#include <iostream>
#include <QCoreApplication>
#include <QFile>
#include <QFileInfo>
#include <QRegularExpression>
#include <QSet>
#include <QStringList>
#include <QTextStream>
#include <QXmlStreamReader>
#include <QXmlStreamWriter>
#include <utility>


char const * inFilename;
QString inFilePath;

QRegularExpression const hasPotentialUnicodeEscape(
    "u[0123456789abcdefABCDEF]{4}");

#define PASS do { writer.writeCurrentToken(reader); } while (false)


QString checkUnicodeEscape(QString const & source,
                           QString const & locationFilename,
                           std::size_t const locationLine)
{
    unsigned here = 0;
    QFile inFile(inFilePath + '/' + locationFilename);
    if (!inFile.open(QIODevice::ReadOnly)) {
        std::cerr << "Failed to open source file: "
                  << locationFilename.toStdString() << std::endl;
        std::exit(5);
    }
    QTextStream in(&inFile);
    for (std::size_t line = 0u; line < locationLine; ++line)
        if (in.readLine().isNull())
            return source;
    std::cout << ++here << std::endl;
    auto line(in.readLine());
    if (line.isNull())
        return source;
    std::cout << ++here << std::endl;
    std::cout << "line = " << line.toStdString() << std::endl;
    line = line.trimmed();
    
    QStringList strings;
    int start = 0;
    for (;;) {
        start = line.indexOf('"', start);
        std::cout << "start = " << start << std::endl;
        if (start < 0)
            break;
        ++start;
        auto end = line.indexOf('"', start);
        std::cout << "end = " << end << std::endl;
        if (end < 0)
            break;
        if (start != end) {
            auto str(line.mid(start, end - start));
            std::cout << "strings add = " << str.toStdString() << std::endl;
            strings.append(std::move(str));
        }
        start = end;
        ++start;
    }
    std::cout << ++here << std::endl;
    
    for (;;) {
        line = in.readLine();
        if (line.isNull())
            break;
        line = line.trimmed();
        if (line.isEmpty())
            continue;
        while (line.startsWith('"')) {
            auto end = line.indexOf('"', 1);
            if (end < 0)
                goto finish;
            if (end > 1) {
                auto str(line.mid(1, end - 1));
                std::cout << "last append = " << str.toStdString() << std::endl;
                strings.last().append(std::move(str));
            }
            line = line.mid(end + 1).trimmed();
        }
        if (!line.isEmpty())
            break;
    }
    
    std::cout << ++here << std::endl;
finish:
    std::cout << ++here << std::endl;

    QSet<QString> rStrings;
    for (auto const & str : strings) {
        std::cout << "check = " << str.toStdString() << std::endl;
        auto s(str);
        s.replace("\\u", "u");
        if (source == s)
            rStrings.insert(str);
    }
    
    switch (rStrings.size()) {
        case 1:
            return *rStrings.begin();
        default:
            std::cerr << "Strange strings in " << locationFilename.toStdString()
                      << ':' << locationLine << std::endl;
            return source;
        case 0:
            return source;
    }
    if (rStrings.size() != 1)
    return source;
    
}

void readError(QXmlStreamReader & reader) {
    std::cerr << inFilename << ':' << reader.lineNumber() << ':'
            << reader.columnNumber() << ": "
            << reader.errorString().toStdString() << std::endl;
    std::exit(4);
}

void recurse(QXmlStreamReader & reader,
             QXmlStreamWriter & writer)
{
    std::size_t mud = 0u;
    QString locationFilename;
    unsigned long long locationLine = 0u;
    QString source;
    while (!reader.atEnd()) {
        switch (reader.readNext()) {
            
        case QXmlStreamReader::Invalid:
            readError(reader);
            
        case QXmlStreamReader::StartElement:
            if (mud) {
                ++mud;
                PASS;
            } else {
                if (reader.qualifiedName() == "location") {
                    auto const attrs(reader.attributes());
                    locationFilename = attrs.value("filename").toString();
                    locationLine = attrs.value("line").toULongLong();
                    ++mud;
                    PASS;
                } else if (reader.qualifiedName() == "source") {
                    PASS;
                    QString source(reader.readElementText());
                    if (source.contains(hasPotentialUnicodeEscape)) {
                        std::cout << "??? " << source.toStdString() << std::endl;
                        source = checkUnicodeEscape(source,
                                                    locationFilename,
                                                    locationLine);
                        std::cout << "--> " << source.toStdString() << std::endl;
                    }
                    writer.writeCharacters(source);
                    writer.writeEndElement();
                } else {
                    ++mud;
                    PASS;
                }
            }
            break;
        case QXmlStreamReader::EndElement:
            if (mud) {
                --mud;
                PASS;
                break;
            } else {
                PASS;
                return;
            }
        default:
            PASS;
            break;
        };
    }
}

template <typename ... Paths>
void recurse(QXmlStreamReader & reader,
             QXmlStreamWriter & writer,
             char const * const path,
             Paths && ... paths)
{
    std::size_t mud = 0u;
    while (!reader.atEnd()) {
        switch (reader.readNext()) {
            
        case QXmlStreamReader::Invalid:
            readError(reader);
            
        case QXmlStreamReader::StartElement:
            if (mud) {
                ++mud;
                PASS;
            } else if (reader.qualifiedName() == path) {
                PASS;
                recurse(reader, writer, paths...);
            } else {
                ++mud;
                PASS;
            }
            break;
        case QXmlStreamReader::EndElement:
            if (mud) {
                --mud;
                PASS;
            } else {
                PASS;
                return;
            }
            break;
        default:
            PASS;
            break;
        };
    }
}

int main(int argc, char * argv[]) {
    QCoreApplication app(argc, argv);
    
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <input file> <output file>"
                  << std::endl;
        return 1;
    }
    
    inFilename = argv[1];
    QFile inFile(inFilename);
    if (!inFile.open(QIODevice::ReadOnly)) {
        std::cerr << "Failed to open input file: " << inFilename << std::endl;
        return 2;
    }
    inFilePath = QFileInfo(inFilename).canonicalPath();
    
    QFile outFile(argv[2]);
    if (!outFile.open(QIODevice::WriteOnly)) {
        std::cerr << "Failed to open output file: " << argv[2] << std::endl;
        return 3;
    }
    
    
    QXmlStreamReader reader(&inFile);
    QXmlStreamWriter writer(&outFile);
    writer.setAutoFormatting(true);
    recurse(reader, writer, "TS", "context", "message");
}

```

**Comment by jaakristioja** [comment]
36dfa22db29df2e67aebcb41af69f9ecf3591b71 Is an improvement of the above code I pushed to master. It takes two arguments: input filename and output filename. Then it lints the input `*.ts` file to the output file, fixing all incorrect `\u` escapes.

In more detail, for all `<source>` strings which match the regular expression `u[0123456789abcdefABCDEF]{4}`, it scans the source code file provided by the `<location>` tag starting from the line number provided by the same `<location>` tag. The line number is the start of the string. It detects all string tokens starting on that line. These strings are then filtered to check if any of those match the source string when all `\u` are replaced with `u`. If and only if exactly one such string exists, this is the one we're looking for. The `<source>` string is then replace with the one in the source file with `\u` escapes properly substituted with the respective Unicode character for that codepoint.

Known issues:
  * Does currently not concatenate string literals on the same line, e.g. `"long" "word"`. But that's more a problem in the source file. This doesn't currently affect us.
  * Does not handle `\U` escapes. This doesn't currently affect us.

@gholmlund Since you're more familiar with the CMake files related to translations, could you please integrate this with `make messages` etc in the master branch and then backport to stable-2.11?

Before we actually push translation file fixes to our repository, we must probably sync with Transifex first and take care we do all this in the correct order so we don't break anything. My guess is that after integrating this fix with our build system (probably on both branchs) we would need to do something like:

1. on stable-2.11 branch: pull all translations from Transifex
2. on stable-2.11 branch: run this tool to patch the translations
3. on stable-2.11 branch: push back the translations with corrected source strings to Transifex
4. on master branch: run this tool to patch the translations
5. on master branch: pull all translations from Transifex

@mgruner Is this order correct? Could you please do this procedure once Gary integrates this workaround?

PS: Feel free to patch and/or modify this tool as needed, or report any problems you find.

**Comment by gholmlund** [comment]
I have added the patcher to cmake for stable-2.11 and master branches. 
@jaakristioja I am not sure the updated escape chars are correct.

**Comment by jaakristioja** [comment]
> I have added the patcher to cmake for stable-2.11 and master branches.

Thanks, @gholmlund !

> @jaakristioja I am not sure the updated escape chars are correct. 

They are not? Can you please be a bit more specific?

**Comment by gholmlund** [comment]
There is an example above in this issue with text of "Itz\u00e1".

After lupdate runs I see this:
    <message>
        <location filename="../../src/backend/managers/clanguagemgr.cpp" line="293"/>
        <source>Itzu00e1</source>
        <extracomment>Language name itz</extracomment>
        <translation>Itzu00e1</translation>
    </message>


After qtbug-35164-patcher runs I see this:
    <message>
        <location filename="../../src/backend/managers/clanguagemgr.cpp" line="293"/>
        <source>Itzu00e1</source>
        <extracomment>Language name itz</extracomment>
        <translation>Itzu00e1</translation>
    </message>

Both are the same. I should expect a added "\", right?

**Comment by jaakristioja** [comment]
@gholmlund Seemingly, it was an off-by-one error which I thought I had fixed, but somehow creeped into master. c9f6e76c99b960502ea892aa9efa20bdb5ddd6c7 fixes this. Could you please retry?

**Comment by gholmlund** [comment]
I see the source is updated, but the translation still has u00e1. Is that correct?
```
    <message>
        <location filename="../../src/backend/managers/clanguagemgr.cpp" line="293"/>
        <source>Itzá</source>
        <extracomment>Language name itz</extracomment>
        <translation>Itzu00e1</translation>
    </message>

```

**Comment by jaakristioja** [comment]
Yes. I don't think we should automatically attempt to fix the translations. We can't even make the tool mark all `<translation>`s where it changes the `<source>` with attribute `type="unfinished"`, because that would break translations on every `make messages` run.

**Comment by mgruner** [comment]
This does not work, unfortunately, despite your impressive work.

If we fix the source strings, we'd have to do that in bibletime_ui.ts, which is what gets pushed to transifex. I tried that (0f057d0d0c0679285ea315a9c6b06faa07c60361), it worked, but then later on a problem appears: whenever "make messages" tries to regenerate the language translation files, it does not find the correct source strings, but instead the broken ones and re-inserts them into the translation file. The proper fix here would have to be done in QT. Alternatively we could opt to no longer regenerate the language .ts files on our own, but leave it to "tx pull", which should have a similar effect.

Another approach here would be to fix just the translation strings instead, *after* pulling them from transifex. That would leave the content in transifex broken, though.

What about removing the offending translation source strings altogether from the sources and not trying to translate them at all? They're just about a few obscure languages AFAICT.

**Comment by jaakristioja** [comment]
> I tried that (0f057d0), it worked, but then later on a problem appears: whenever "make messages" tries to regenerate the language translation files, it does not find the correct source strings, but instead the broken ones and re-inserts them into the translation file.

Ah, ok. In general, 0f057d0 seems to work as I thought it was supposed to. But you're correct to note that successive `make messages` calls break the translations. To be more precise, it seems that every next invocation of `lupdate` scans the C++ source files and detects the source strings incorrectly again (i.e. it detects "Itzu00e1"). Then it looks at our fixed `.ts` file and doesn't find `<source>Itzu00e1</source>` (because it now correctly contains `<source>Itzá</source>`). Therefore it inserts a new untranslated `<message>` with `<source>Itzu00e1</source>` and marks `<translation>` of the non-detected `<message>` with `<source>Itzá</source>` as `<translation type="vanished">`.

I think that a simple workaround would be to re-break the translations (i.e. undo the work of `qtbug-35164-patcher`) before each `lupdate` invocation. More specifically all non-ASCII characters in `<source>` strings must be converted to broken unicode escapes, e.g. `á` must again become `u00e1` (without the backslash). I think I can easily implement an `--undo` mode in `qtbug-35164-patcher`.

**Comment by mgruner** [comment]
I would like to ask if that’s actually needed. If we skip lupdate calls for the language .ts files, and leave that to »tx pull«, it should be fine. It pulls the translations based on the previously pushed latest state of the bibletime_ui.ts file.

**Comment by jaakristioja** [comment]
[QTBUG-35164](https://bugreports.qt.io/browse/QTBUG-35164) was fixed for Qt 5.9 which is currently our minimum required version. All the buggy `<source>Itzu00e1</source>` stuff will be fixed when `make messages` is run. The workaround we had was already removed from tree in commit a66d672164ba74517c4bd92fc9f54e6edf17c855. Closing.
